Scheduler for regular search evaluation runs

ajleong623 opened the pull request:
### Description
For requesting a regularly scheduled search evaluation, the user could add an `cron` parameter to denote the cron job schedule for running search evaluation. 

Some changes that are made are that there are now 3 new APIs for interacting with scheduling experiments. The endpoints are `experiment/<job_id>/schedule` which is applied to the GET and DELETE methods and `experiment/schedule` which is applied to the GET and POST methods. 

There are 2 new indices, `.scheduled-jobs` and `search-relevance-scheduled-experiment-history`. The purpose of the `.scheduled-jobs` index is to store the currently running experiment schedules. The `search-relevance-scheduled-experiment-history` index stores the historical experiment results with timestamps which were resulted from the scheduled job runner. 

Unit and integration tests are provided, however, additions such as workload management, integration with alerting and resource monitoring are not available in this pull request, but I would like to add those into a future pull request. 

Please let me know if there are any questions or concerns. 

### Issues Resolved
#213 #226 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

epugh commented on src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java:
```diff
@@ -26,6 +26,8 @@ private PluginConstants() {}
     /** The URI for this plugin's search configurations rest actions */
     public static final String SEARCH_CONFIGURATIONS_URL = SEARCH_RELEVANCE_BASE_URI + "/search_configurations";
+    /** The URI for this plugin's search configurations rest actions */
+    public static final String SCHEDULED_EXPERIMENT_URL = SEARCH_RELEVANCE_BASE_URI + "/scheduled_experiment";
```
could also have /experiments/scheduled to get all the jobs...

epugh commented on src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java:
```diff
@@ -80,8 +84,10 @@ private PluginConstants() {}
     public static final String SEARCH_PIPELINE = "searchPipeline";
     public static final String SIZE = "size";
     public static final String QUERYSET_ID = "querySetId";
+    public static final String EXPERIMENT_ID = "experiment_id";
```
don't mix camels and snakes!

epugh commented on src/main/java/org/opensearch/searchrelevance/dao/ScheduledJobsDao.java:
```diff
@@ -0,0 +1,124 @@
+     * Create scheduled jobs index if not exists
+     * @param stepListener - step lister for async operation
+     */
+    public void createIndexIfAbsent(final StepListener<Void> stepListener) {
```
again, this is the suss thing.. shouldn't the job scheduler own this?

ajleong623 replied:
As discussed above, I think the index is owned by this plugin, but the job scheduler only listens to the index based on the name.

epugh commented on src/main/java/org/opensearch/searchrelevance/plugin/SearchRelevancePlugin.java:
```diff
@@ -172,6 +200,15 @@ public Collection<Object> createComponents(
+        jobRunner.setQuerySetDao(querySetDao);
+        jobRunner.setSearchConfigurationDao(searchConfigurationDao);
+        jobRunner.setMetricsHelper(metricsHelper);
+        jobRunner.setHybridOptimizerExperimentProcessor(new HybridOptimizerExperimentProcessor(judgmentDao, experimentTaskManager));
```
we aren't trying to schedule hybrid jobs, so not sure about the above line...   THough, how easy would it be to schedule a hybrid job?

epugh replied:
actually, since a hybrid job is stil lan experiment,t hen we can leave this!

epugh commented on src/main/java/org/opensearch/searchrelevance/plugin/SearchRelevancePlugin.java:
```diff
@@ -233,10 +274,85 @@ public List<RestHandler> getRestHandlers(
 
+    @Override
+    public String getJobType() {
+        return "scheduler_sample_extension";
```
???

epugh commented on src/main/java/org/opensearch/searchrelevance/rest/RestDeleteScheduledExperimentAction.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public List<Route> routes() {
+        return singletonList(new Route(DELETE, String.format(Locale.ROOT, "%s/{%s}", SCHEDULED_EXPERIMENT_URL, DOCUMENT_ID)));
```
DELETE /experiment/{experiment_id}/schedule.  No need for job id since you can ONLY have one job per experiment.  You can't run an experiment nightly and on the hour.   That is part of our security/validation layer!  Even though under the covers it could be done...

epugh commented on src/main/java/org/opensearch/searchrelevance/rest/RestPostScheduledExperimentAction.java:
```diff
@@ -0,0 +1,93 @@
+
+@Log4j2
+/**
+ * Rest Action to facilitate requests to create a scheduled experiment.
```
```suggestion
 * Rest Action to facilitate requests to schedule running an experiment.
```

epugh commented on src/main/java/org/opensearch/searchrelevance/rest/RestDeleteScheduledExperimentAction.java:
```diff
@@ -0,0 +1,98 @@
+
+@Log4j2
+/**
+ * Rest Action to facilitate requests to delete a scheduled experiment.
```
```suggestion
 * Rest Action to facilitate requests to delete the scheduled running of an experiment```

epugh commented on src/main/java/org/opensearch/searchrelevance/rest/RestPutExperimentAction.java:
```diff
@@ -47,6 +47,7 @@
 @AllArgsConstructor
 public class RestPutExperimentAction extends BaseRestHandler {
     private static final String PUT_EXPERIMENT_ACTION = "put_experiment_action";
+    static final String JOB_INDEX_NAME = ".scheduler_sample_extension";
```
???

epugh commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,469 @@
+    public void runJob(ScheduledJobParameter jobParameter, JobExecutionContext context) {
+        if (!(jobParameter instanceof SearchRelevanceJobParameters)) {
+            throw new IllegalStateException(
+                "Job parameter is not instance of SampleJobParameter, type: " + jobParameter.getClass().getCanonicalName()
```
```suggestion
                "Job parameter is not instance of SearchRelevanceJobParameters, type: " + jobParameter.getClass().getCanonicalName()
```

epugh commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,469 @@
+                    SearchRelevanceJobParameters parameter = (SearchRelevanceJobParameters) jobParameter;
+                    log.info(parameter.getExperimentId());
+                    String experimentId = parameter.getExperimentId();
+                    try {
```
I suspect everything in this try can be refactored out of where else we run an experiment, so we don't duplicate all the logic....

ajleong623 replied:
I think this part was resolved by Eric, but I will probably make a separate class for the logic anyways and remove the comment.

epugh reviewed (dismissed):
Progress!   We are now on the cron pattern.  Now to think about nesting the API under the /experiment/{experiment_id}/schedule name space.

cwperks commented on src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java:
```diff
@@ -53,6 +55,8 @@ private PluginConstants() {}
     public static final String JUDGMENT_CACHE_INDEX_MAPPING = "mappings/judgment_cache.json";
     public static final String EXPERIMENT_VARIANT_INDEX = "search-relevance-experiment-variant";
     public static final String EXPERIMENT_VARIANT_INDEX_MAPPING = "mappings/experiment_variant.json";
+    public static final String SCHEDULED_JOBS_INDEX = ".scheduled-jobs";
```
job scheduler has an index listener where it listens to the indices from this repo's implementation of `JobSchedulerExtension` so a plugin still needs to define the index and do index operations into this index.

That being said, this index name is too generic. It should be more specific to this plugin's use-case.

ajleong623 commented:
I believe I have addressed the comments. One of them, I did add a TODO comment so that it can be addressed in the future. Right now, the solution to refactoring the logic of running experiments is a bit involved.

epugh commented:
This PR is ready for review it looks like!

ajleong623 commented:
@martin-gaievski Thank you so much for your review and taking the time to look through the code. I know that the PR is a bit long and can be hairy. 

For rate limiting the amount of scheduled jobs, the minimum possible interval for the cron job is once every minute. I am not completely sure how many jobs it takes to cause degradation or how to measure it, but for now, I am thinking of just having 3 jobs run at once and keeping an atomic object in a singleton class as a naive solution. 

I am also reading up a little bit on what concurrency techniques might be available to me, but I will aim to update the code later on today so that everything will be addressed by tomorrow.

fen-qin commented on src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java:
```diff
@@ -26,6 +26,11 @@ private PluginConstants() {}
+    /** The URI for this plugin's scheduled experiments rest actions. For retrieving/deleting individual
+     * experiments, the experimentId is specified as a path parameter in between the prefix and suffix.
+    */
+    public static final String SCHEDULED_EXPERIMENT_URL_PREFIX = SEARCH_RELEVANCE_BASE_URI + "/experiment";
```
we had a path `EXPERIMENTS_URI = SEARCH_RELEVANCE_BASE_URI + "/experiments"` for experiment API, do you think we shall combine both, otherwise it will be little bit confuse on /experiment and /experiments

fen-qin commented on src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java:
```diff
@@ -53,6 +58,10 @@ private PluginConstants() {}
     public static final String JUDGMENT_CACHE_INDEX_MAPPING = "mappings/judgment_cache.json";
     public static final String EXPERIMENT_VARIANT_INDEX = "search-relevance-experiment-variant";
     public static final String EXPERIMENT_VARIANT_INDEX_MAPPING = "mappings/experiment_variant.json";
+    public static final String SCHEDULED_JOBS_INDEX = ".scheduled-jobs";
```
why to use system index here ? does it store sensitive data ? or is it expected not to be visible to dashboards in the future ?

fen-qin commented on src/main/java/org/opensearch/searchrelevance/rest/RestPostScheduledExperimentAction.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        if (!settingsAccessor.isWorkbenchEnabled()) {
```
we can use another flag to block new features specificly

fen-qin commented on src/main/java/org/opensearch/searchrelevance/utils/CronUtils.java:
```diff
@@ -0,0 +1,53 @@
+            return new CronValidationResult(false, "cron expression cannot be null or empty");
+        }
+
+        CronParser cronParser = new CronParser(CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX));
```
please add more validation, like MIN, MAX cron that will be allowed for customers to input. Also, probably it's better for cron schedule status or error message to be returned back to customers.

ajleong623 replied:
For the cron schedule status, would I make a new exception class?

fen-qin commented on src/main/java/org/opensearch/searchrelevance/dao/ScheduledExperimentHistoryDao.java:
```diff
@@ -0,0 +1,126 @@
+ */
+package org.opensearch.searchrelevance.dao;
+
+import static org.opensearch.searchrelevance.indices.SearchRelevanceIndices.SCHEDULED_EXPERIMENT_HISTORY;
```
the scheduled experiment history may grow quickly. please ass retention policies. Same for scheduled dao, if the definition for scheduled job is for any running or pending jobs, then we should do proper clean up after jobs are completed.

fen-qin replied:
is there any use case for history data ? any query or list heavy operations ? if no, then you can ignore.
But for scheduled jobs, it might become an issue when a job is stuck in either 'ENABLED' or 'PENDING' status, and it will continuously be picked up by the scheduler each time it runs, so I'm wondering if we should implement a cleanup mechanism. Probably with a new 'TIMEOUT' status to handle these problematic stuck jobs.

fen-qin commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,550 @@
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.client.Client;
+
+public class SearchRelevanceJobRunner implements ScheduledJobRunner {
```
please consider multi-thread use case. If two threads simultaneously update, for example, setJobName or setLastUpdateTime, it can lead to race conditions

wrigleyDan reviewed (commented):
Additional questions to those in the review comments:

Looking at the documents in ``
```
{
        "_index": "search-relevance-scheduled-experiment-history",
        "_id": "66829d27-83cf-4a66-b8af-eff0b6483645",
        "_score": 1,
        "_source": {
          "id": "66829d27-83cf-4a66-b8af-eff0b6483645",
          "experimentId": "e7ac2d44-0290-4ac5-a85e-c7dc3b481871",
          "jobId": "e7ac2d44-0290-4ac5-a85e-c7dc3b481871",
          "jobIndexName": "index",
          "timestamp": "2025-09-04T06:59:00.972Z",
          "status": "COMPLETED",
          "results": [
            {
              "evaluationId": "e0e1231c-4e90-46b4-b435-01786ab5db2f",
              "queryText": "tv",
              "searchConfigurationId": "047f382c-676a-456b-9374-84d73c1a9a24"
            },
            {
              "evaluationId": "f866f6b9-148b-4869-af83-7faa34bf3ba9",
              "queryText": "led tv",
              "searchConfigurationId": "047f382c-676a-456b-9374-84d73c1a9a24"
            }
          ]
        }
      }
```
The value for the field `jobIndexName` appears to be `index`. What is the purpose of that field? If there is no purpose I suggest we get rid off it. If there is a purpose we need meaningful values for it.

In the following I try to sum up some data model considerations since there has been some discussion during our planning meeting on Wednesday and on Slack.
My understanding is:
* Currently, creating a scheduler for an experiment results in a job created in the index `.scheduled-jobs` (that needs a better name btw).
* For each cron iteration (for example, every 24 hours) an experiment is triggered. After completion this results in a new entry in the index `search-relevance-scheduled-experiment-history`. This is a new index.
* Every evaluation of this experiment (this means for every query) the evaluation result data is stored in the index `search-relevance-evaluation-result`. This is an already existing index.

In the UI, all experiments are listed in a tabular overview. These experiments are retrieved from the `.plugins-search-relevance-experiment` index. This means that scheduled experiments are "invisible" to the UI right now.

To enable consistent searching, sorting, paging through experiments in the overview I think we should store all experiments in one index. I don't think they need to be that different that it justifies separate indexes.

I see two possible paths forward: 
1) We treat every run of a scheduled experiment as an individual experiment. We remove the `search-relevance-scheduled-experiment-history` index and store them in `.plugins-search-relevance-experiment`. Tracking the performance of scheduled experiments over time (how does my search quality evolve over time) is a requirement, so this would require adding something like a `scheduledExperimentId` to the individual experiment runs. This could "pollute" the tabular overview. 
2) We treat scheduled experiment similarly to hybrid search optimization experiments. There is one "parent" experiment, and several "child" experiments. At the moment we call the individual hybrid search optimization experiments "variants" as they are variants in the parameter space. I don't think "variant" is the right notation for scheduled experiments, though. Additionally, the use case to schedule hybrid search optimization experiments should be possible for users, so again, a new field like `scheduledExperimentId` is needed to group scheduled experiment runs to the scheduled job.

What I don't know: Does adding a new field to an existing index mess up backwards compatibility?

cwperks commented on docs/feature-design/regularly-scheduled-experiments.md:
```diff
@@ -0,0 +1,350 @@
+- New APIs are added which means that the specification should be updated.
+
+
+## Security Considerations
```
FYI Security for scheduled jobs in less then idea in OpenSearch (at least atm). Currently, the paradigm used is called [Roles Injection](https://github.com/opensearch-project/common-utils/blob/main/src/main/java/org/opensearch/commons/InjectSecurity.java#L170-L188) which requires the plugin that schedules job to inject roles into the threadcontext at job runtime. This requires the plugin to store a copy of the user at the time of job creation as part of the job metadata. 

I'm aiming to change the paradigm to remove Roles Injection in the future (See [RFC for platform-level security](https://github.com/opensearch-project/security/issues/5052)) to a token-based model where the token is passed at job creation and has only the permissions it needs for the job to run. 

In absence of roles injection, the job runner runs will system-level permissions and can do anything. Remember, jobs are not like handling API Requests where there are credentials to authenticate and hence the issue.

heemin32 replied:
Thanks for the valuable insights from a security perspective. For now, even without scheduling, the experiment is secured at the API level, and we don’t use role injection to run experiments with user credentials. We can consider this as a follow-up item if there’s a strong need for role injection in experiments.

ajleong623 commented on src/main/java/org/opensearch/searchrelevance/indices/SearchRelevanceIndices.java:
```diff
@@ -66,7 +70,17 @@ public enum SearchRelevanceIndices {
      * Experiment Variant Index
      */
-    EXPERIMENT_VARIANT(EXPERIMENT_VARIANT_INDEX, EXPERIMENT_VARIANT_INDEX_MAPPING, false);
+    EXPERIMENT_VARIANT(EXPERIMENT_VARIANT_INDEX, EXPERIMENT_VARIANT_INDEX_MAPPING, false),
```
I added a couple more search relevance indices. The semicolon is at the last index I added.

ajleong623 commented on src/main/java/org/opensearch/searchrelevance/transport/scheduledJob/PostScheduledExperimentTransportAction.java:
```diff
@@ -0,0 +1,96 @@
+            String id = experimentId; // Since there is at most 1 scheduled job per experiment, the ids could be the same.
+            SearchRelevanceJobParameters jobParameter = new SearchRelevanceJobParameters(
+                id,
+                "experiment-parameters",
```
It is not needed.

ajleong623 commented:
@wrigleyDan @cwperks Thank you for taking a look into the pull request and for your extensive reviews. To answer some general questions, @wrigleyDan I found that adding the field into the index will not break backwards compatibility as the mapping for `evaluation_result` has actually been changed in the past in https://github.com/opensearch-project/search-relevance/pull/174.

ajleong623 commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,550 @@
+    ) {
+        Map<String, SearchConfigurationDetails> searchConfigurations = new HashMap<>();
+        AtomicInteger pendingConfigs = new AtomicInteger(request.getSearchConfigurationList().size());
+        AtomicBoolean hasFailure = new AtomicBoolean(false);
```
I added an attempt to convert the logic to using futures. I created the list of futures so that they could all be completed asynchronously and then added the `searchConfigurations` afterwards so that the synchronization block around the `searchConfigurations` would not be necessary.

ajleong623 commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,550 @@
+        // TODO: A lot of the logic here is reused from PutTransportExperiment.
+        // Eventually we have to abstract it in another class to reduce complexity.
+
+        Runnable runnable = () -> {
```
I added a timeout mechanism, but I am not sure about some details. If the scheduled experiment is cancelled, due to timeout, it is clean when the task to run that experiment has not been run yet. However, should the experiment be cancelled while it is running. I imagine this part becomes much more complex. Additionally, which cleanup actions were you thinking about. I do not think anything is initialized when the task is submitted to executor for the thread pool, however, if the task has started running, there might be some resources.

ajleong623 commented:
Other than the strategy for updates to the mappings, the concurrency and timeout framework, and including information on the security roles, I believe all comments of the PR have been addressed. 

One part that might need more explanation is how the indices interact. The 3 relevant indices for this PR are `search-relevance-evaluation-result`, `search-relevance-scheduled-experiment-history`, and `.search-relevance-scheduled-experiment-jobs`.

The models: 
`search-relevance-evaluation-result`

```
  private final String id;
  private final String timestamp;
  private final String searchConfigurationId;
  private final String searchText;
  private final List<String> judgmentIds;
  private final List<String> documentIds;
  private final List<Map<String, Object>> metrics;
  private final String experimentId;
  private final String experimentVariantId;
  private final String experimentVariantParameters;
  private final String scheduledExperimentId;
```

`search-relevance-scheduled-experiment-history`

```
  private final String id;
  private final String experimentId;
  private final String timestamp;
  private final AsyncStatus status;
  private final List<Map<String, Object>> results;
```

`.search-relevance-scheduled-experiment-jobs`

```
  private final String id;
  private final Instant lastUpdateTime;
  private final Instant enabledTime;
  private final boolean isEnabled;
  private final Schedule schedule;
```

```
mermaid
sequenceDiagram
    PutExperimentTransportAction/ScheduledExperimentRunnerManager->>+Hybrid/Pointwise ExperimentProcessor: Starts running a pointwise or hybrid search experiment
    Hybrid/Pointwise ExperimentProcessor->>+search-relevance-evaluation-result: Submits evaluation_result
```

epugh commented:
I wonder if describing the indexes in DEVELOPER_GUIDE would be a good addition?   YOu could move them into that file as part of this PR, and we could use that to talk about it?  There is enough complexity here that the next person will need to understand this!

ajleong623 commented on build.gradle:
```diff
@@ -195,6 +206,8 @@ dependencies {
     api "org.opensearch:opensearch:${opensearch_version}"
     api group: 'org.opensearch', name:'opensearch-ml-client', version: "${opensearch_build}"
+    compileOnly "org.opensearch:opensearch-job-scheduler-spi:${opensearch_build}"
+    compileOnly "org.opensearch:opensearch-job-scheduler:${opensearch_build}"
```
Removed.

epugh commented:
Tried some things out...   @ajleong623 I shared a patch file to get tests to compile (and some pass again!) on Slack.

What I think i am seeing is that the `scheduledExperimentId` remains the same for each run (and duplicates the `experimentId`), and the `experimentVariantId` is changing per run.   We need the `scheduledExperimentId` to be the same value per execution loop across ALL the queries so that we can power our visualization graph.   

It's cool to dig into how dashboard visualizations work!

ajleong623 commented:
I also changed the api endpoints so now when deleting or retrieving a job, the `experiment_id` parameter is at the end of the url path. I did this because having the parameter in the middle felt a little bit awkward, and also did not fit the pattern of the other endpoints.

ajleong623 commented:
Just a personal reminder to add documentation comments throughout code changes before submitting.

fen-qin commented on src/main/java/org/opensearch/searchrelevance/utils/ConcurrencyUtil.java:
```diff
@@ -0,0 +1,39 @@
+
+        ScheduledFuture<?> timeout = threadPool.scheduler().schedule(() -> {
+            if (timeoutFuture.cancel(false)) {
+                future.cancel(true);
```
please resolve the build failure
```
Forbidden method invocation: java.util.concurrent.Future#cancel(boolean) [Don't interrupt threads use FutureUtils#cancel(Future<T>) instead]
```

sample code:
```
    // Before:
    // future.cancel(true);

    // After:
    FutureUtils.cancel(future);
```

martin-gaievski commented on src/main/java/org/opensearch/searchrelevance/executors/ExperimentRunningManager.java:
```diff
@@ -0,0 +1,601 @@
+    ) {
+        List<Future<?>> futures = new ArrayList<>();
+        if (request.getScheduledExperimentResultId() != null) {
+            runningFutures.put(request.getScheduledExperimentResultId(), futures);
```
with such logic for adding items into the map, the values are not thread-safe, multiple threads can mutate same list concurrently. 
You can do something like this when you're adding to map:
```
runningFutures.compute(scheduledExperimentResultId, (key, existingList) -> {
    List<Future<?>> list = existingList != null ? existingList : Collections.synchronizedList(new ArrayList<>());
    list.add(future);
    return list;
});
```

martin-gaievski commented on src/main/java/org/opensearch/searchrelevance/executors/ExperimentRunningManager.java:
```diff
@@ -0,0 +1,601 @@
+            request,
+            searchConfigurations,
+            queryTextWithReferences,
+            finalResults,
```
looks like we're simply accumulating results into this List of maps `finalResults`. This could consume significant memory. Let's at least signal to the logs in case we reached certain high threshold, probably Runtime class can give some helpful info: `Runtime.getRuntime().totalMemory()` or `Runtime.getRuntime().freeMemory()`

martin-gaievski commented on src/main/java/org/opensearch/searchrelevance/executors/ExperimentRunningManager.java:
```diff
@@ -0,0 +1,601 @@
+                return;
+            }
+            if (checkIfCancelled(cancellationToken)) {
+                log.info("Experiment has been timed out while executing experiments for each queryText");
```
can you add more details, e.g.
```log.info("Experiment {} timed out after {}ms while processing query '{}' ({}/{} completed)", 
         experimentId, elapsedTime, queryText, completed, total);```

martin-gaievski reviewed (commented):
Added some suggestions, few I would say are need to be addressed, rest is up to you, can be done in a follow up PR:
- concurrency and thread safety of futures in the concurrent map in ExperimentRunningManager
- resource leakage in SearchRelevanceJobRunner and memory in ExperimentRunningManager

ajleong623 commented on src/main/java/org/opensearch/searchrelevance/scheduler/SearchRelevanceJobRunner.java:
```diff
@@ -0,0 +1,138 @@
+            log.error("Timeout for scheduled experiment has occured!");
+        } catch (CompletionException e) {
+            log.error("Scheduled experiment has timed out. Moving onto cleanup");
+        } finally {
```
I might be making a bit of an assumption, but I think that all the scheduled tasks should eventually either finish as a failure (In which case it would run https://github.com/opensearch-project/search-relevance/pull/220/files#diff-619c58b6692abd9bd3d4704b7c573cb31bacfe94a5279707b727d54ec9bfebe4R575) or a success (https://github.com/opensearch-project/search-relevance/pull/220/files#diff-619c58b6692abd9bd3d4704b7c573cb31bacfe94a5279707b727d54ec9bfebe4R527)

martin-gaievski replied:
I agree this argument is valid for a normal happy path case, in such case latch will be decremented.

But what happens if the experiment task never starts or is interrupted before reaching those methods? I can think of following scenarios:

1. Thread pool rejection
```java
// If thread pool is saturated:
CompletableFuture.runAsync(jobRunTask, threadPool.generic())
// throws RejectedExecutionException BEFORE task starts
// jobRunTask never executes → countDown() never called
```

2. Interruption during join():
```java
searchEvaluationTask.join(); // Thread interrupted here
// throws CompletionException wrapping InterruptedException
// task may be cancelled before completion
```

3. Timeout wrapper failure:
```java
ConcurrencyUtil.withTimeout(...) // Internal exception here
// task setup fails before execution
// experiment never starts → no countDown()
```

4. Early cancellation:
```java
if (cancellationToken.isCancelled()) // Cancelled during setup
// task exits early without reaching final methods
```

This is how this impacts system:
```
try {
    searchEvaluationTask.join(); // exception here
} catch (CancellationException e) {
    log.error("Timeout occurred!");
    // latch not decremented - leak
} catch (CompletionException e) {
    log.error("Task failed!");  
    // latch not decremented - leak
} finally {
    // cleanup runs but latch still at count 1
    manager.cleanupResources(...);
    // any thread waiting on actuallyFinished.await() blocks forever
}
```

martin-gaievski reviewed (commented):
Few important finding are not addressed, please review my comments

ajleong623 commented:
@martin-gaievski Outside of the memory logging (I have made a separate issue for this), I have incorporated your new comments. I also added some new tests based on the concurrency issues mentioned. 

Latch Cleanup: Covered in `ScheduledExperimentRunnerManagerTests.java`, `SearchRelevanceJobRunnerTests.java`

Concurrent access to futures: Covered in `ExperimentRunningManagerTests.java`

Cancellation of experiment: `ScheduledExperimentRunnerManagerTests.java`, `ExperimentRunningManagerTests.java`, `HybridOptimizerExperimentProcessorTests.java`, `ExperimentTaskManagerTests.java`. 

Happy path: `ScheduledJobsIT.java`

Let me know if you have any questions or concerns or if there is anything I missed.

epugh reviewed (approved):
A long road to get here, but an awesome feature!
